### PR TITLE
LPS-109844 Create drafts of structured content

### PR DESCRIPTION
**Pull request by LuismiBarcos**

### CODE REVIEW

File: modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dynamic/data/mapping/DDMFormFieldUtil.java [hunk 1]
```diff
@@ -12,7 +12,7 @@
  * details.
  */
 
-package com.liferay.headless.delivery.internal.dto.v1_0.util;
```
Comment (nhpatt) [inline modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dynamic/data/mapping/DDMFormFieldUtil.java]: Doing the 2 step (move and adapt) would have moved this changes to the adapt and separate from exposing in the bnd file that could be their own commit because you are changing different things, java and bnds.

File: modules/apps/headless/headless-admin-content/headless-admin-content-impl/src/main/java/com/liferay/headless/admin/content/internal/resource/v1_0/StructuredContentResourceImpl.java [hunk 2]
```diff
@@ -197,6 +230,90 @@ public EntityModel getEntityModel(MultivaluedMap multivaluedMap) {
...
 
+	@Override
+	public StructuredContent postSiteStructuredContentDraft(
+			@NotNull Long siteId, StructuredContent structuredContent)
+		throws Exception {
+
+		DDMStructure ddmStructure = DDMStructureUtil.getDDMStructure(
+			_ddmStructureService, structuredContent);
+
+		LocalDateTime localDateTime = LocalDateTimeUtil.toLocalDateTime(
```
Comment (nhpatt) [inline modules/apps/headless/headless-admin-content/headless-admin-content-impl/src/main/java/com/liferay/headless/admin/content/internal/resource/v1_0/StructuredContentResourceImpl.java:241]: Variables are defined in the order they are used and you don't need localDateTime here, I think this should be right before servicecontext

File: modules/apps/headless/headless-admin-content/headless-admin-content-impl/src/main/java/com/liferay/headless/admin/content/internal/resource/v1_0/StructuredContentResourceImpl.java [hunk 3]
```diff
@@ -197,6 +230,90 @@ public EntityModel getEntityModel(MultivaluedMap multivaluedMap) {
...
+	}
+
+	private String _getDDMTemplateKey(DDMStructure ddmStructure) {
+		List<DDMTemplate> ddmTemplates = ddmStructure.getTemplates();
+
+		if (ddmTemplates.isEmpty()) {
+			return StringPool.BLANK;
+		}
+
+		DDMTemplate ddmTemplate = ddmTemplates.get(0);
```
Comment (nhpatt) [inline modules/apps/headless/headless-admin-content/headless-admin-content-impl/src/main/java/com/liferay/headless/admin/content/internal/resource/v1_0/StructuredContentResourceImpl.java:312]: I guess this is my code. It's awful but I don't know how to improve it...

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormValuesUtil.java [hunk 4]
```diff
@@ -79,6 +84,46 @@ public static DDMFormValues toDDMFormValues(
 		};
 	}
 
+	public static String toText(
```
Comment (nhpatt) [inline modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormValuesUtil.java:88]: toText seems to abstract, could it be toContent or something like that?

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMStructureUtil.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public class DDMStructureUtil {
+
+	public static DDMStructure getDDMStructure(
+			DDMStructureService ddmStructureService,
+			StructuredContent structuredContent)
+		throws Exception {
+
+		try {
+			return ddmStructureService.getStructure(
```
Comment (nhpatt) [inline modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMStructureUtil.java]: mmm encapsulating this seems to much. Ideally it would be a exceptcionMapper but I don't think it can be done because the exception is too generic. I would leave this way but I don't know if Brian is going to change it.

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/StructuredContentUtil.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+				ddmStructure,
+				ddm.getFields(ddmStructure.getStructureId(), serviceContext),
+				ddmStructure.getGroupId());
+		}
+		finally {
+			LocaleThreadLocal.setSiteDefaultLocale(originalSiteDefaultLocale);
+		}
+	}
+
+	public static Map<String, Serializable> getExpandoBridgeAttributes(
```
Comment (nhpatt) [inline modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/StructuredContentUtil.java]: Too little gain for encapsulating this, right?

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormFieldUtil.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+				if (_log.isWarnEnabled()) {
+					_log.warn(portalException, portalException);
+				}
+			}
+		}
+
+		return null;
+	}
+
+	public static List<DDMFormField> getRootDDMFormFields(
```
Comment (nhpatt) [inline modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormFieldUtil.java]: this too, I prefer a smaller API footprint that repeat one/two lines

**Review by nhpatt — COMMENTED** [review]
Let's wait for Javier's review.

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormValuesUtil.java (hunk 4 above)
Comment (javierdearcos) [reply modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormValuesUtil.java:88]: I agree. Or maybe serializeContent

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormValuesUtil.java [hunk 8]
```diff
@@ -79,6 +84,46 @@ public static DDMFormValues toDDMFormValues(
...
+				ddmFormValues);
+
+		DDMFormValuesSerializerSerializeResponse
+			ddmFormValuesSerializerSerializeResponse =
+				jsonDDMFormValuesSerializer.serialize(builder.build());
+
+		return ddmFormValuesSerializerSerializeResponse.getContent();
+	}
+
+	public static void validateDDMFormValues(
```
Comment (javierdearcos) [inline modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormValuesUtil.java:111]: Can we extract this behavior to a ExceptionMapper and avoid this method? Just call ddmFormValuesValidator.validate from the impl modules

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMStructureUtil.java (hunk 5 above)
Comment (javierdearcos) [reply modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMStructureUtil.java]: We already have a PrincipalExceptionMapper in Vulcan, that maps to NOT FOUND in get requests and to FORBIDDEN in other cases. Maybe we can use it, so we have a uniform behavior with other endpoints

https://github.com/liferay/liferay-portal/blob/master/modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/exception/mapper/PrincipalExceptionMapper.java

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/StructuredContentResourceImpl.java [hunk 9]
```diff
@@ -16,9 +16,6 @@
 
 import com.liferay.document.library.kernel.service.DLAppService;
 import com.liferay.dynamic.data.mapping.io.DDMFormValuesSerializer;
```
Comment (javierdearcos) [inline modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/StructuredContentResourceImpl.java:18]: I think this is a good smell. A refactor that allow you to get rid of a lot of imports

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/StructuredContentResourceImpl.java [hunk 10]
```diff
@@ -16,9 +16,6 @@
 
 import com.liferay.document.library.kernel.service.DLAppService;
```
Comment (javierdearcos) [inline modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/StructuredContentResourceImpl.java:17]: I think this is a good smell. A refactor that allow you to get rid of a lot of imports

File: modules/apps/headless/headless-delivery/headless-delivery-api/bnd.bnd [hunk 11]
```diff
@@ -3,7 +3,9 @@ Bundle-SymbolicName: com.liferay.headless.delivery.api
...
 Export-Package:\
 	com.liferay.headless.delivery.dto.v1_0,\
+	com.liferay.headless.delivery.dynamic.data.mapping,\
 	com.liferay.headless.delivery.resource.v1_0,\
 	com.liferay.headless.delivery.search.aggregation,\
 	com.liferay.headless.delivery.search.filter,\
-	com.liferay.headless.delivery.search.sort
\ No newline at end of file
+	com.liferay.headless.delivery.search.sort,\
+	com.liferay.headless.delivery.util
```
Comment (javierdearcos) [inline modules/apps/headless/headless-delivery/headless-delivery-api/bnd.bnd:11]: Don't know if it is a little bit generic. We can try this way for now. Other option would be `com.liferay.headless.delivery.dto.v1_0.util` (It is possible that if we change the dtos in a v2 of the apis all this classes become useless)

**Review by javierdearcos — COMMENTED** [review]
LGTM

File: modules/apps/headless/headless-admin-content/headless-admin-content-impl/src/main/java/com/liferay/headless/admin/content/internal/resource/v1_0/StructuredContentResourceImpl.java (hunk 2 above)
Comment (LuismiBarcos) [reply modules/apps/headless/headless-admin-content/headless-admin-content-impl/src/main/java/com/liferay/headless/admin/content/internal/resource/v1_0/StructuredContentResourceImpl.java]: Done!

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormValuesUtil.java (hunk 4 above)
Comment (LuismiBarcos) [reply modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormValuesUtil.java]: I renamed it to `serializeContent`

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMStructureUtil.java (hunk 5 above)
Comment (LuismiBarcos) [reply modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMStructureUtil.java]: Finally, I deleted the DDMStructureUtil and use the Principal ExceptionMapper. I was looking for the reason because we have this in this way but I didn't find anything. Just [this LPS](https://issues.liferay.com/browse/LPS-93617) where was added.

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormFieldUtil.java (hunk 7 above)
Comment (LuismiBarcos) [reply modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormFieldUtil.java]: Done!

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormValuesUtil.java (hunk 8 above)
Comment (LuismiBarcos) [reply modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/DDMFormValuesUtil.java]: Done!

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/StructuredContentUtil.java (hunk 6 above)
Comment (LuismiBarcos) [reply modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/dto/v1_0/util/StructuredContentUtil.java]: Yes, I changed it

**Review by nhpatt — APPROVED** [review]

**Comment by nhpatt** [comment]
ci:test:relevant

**Review by javierdearcos — APPROVED** [review]
👏

**Comment by javierdearcos** [comment]
My only remaining doubt is if the DDMFormValuesValidationExceptionMapper will be triggered from the Headless Admin Content endpoint.

Making a quick search I think you can add the two applications to the exception mapper like this:
```
osgi.jaxrs.application.select=(!(osgi.jaxrs.name=Liferay.Headless.Delivery)(osgi.jaxrs.name=Liferay.Headless.Admin.Content))
```
Although is should be tested as I'm not sure it is exactly like this.

Other option will be just to duplicate it in the Headless Admin Content project speficying this application in the  `osgi.jaxrs.application.select` property

**Comment by LuismiBarcos** [comment]
I was not able to do like @javierdearcos said. I tried other options but none works 😔
So, I duplicated it in Headless Admin Content. Thinking about it maybe is the best for now because specify that we select the application `Headless Admin Content` in an exception mapper inside `Headless Delivery` could be confused, right?

**Comment by nhpatt** [comment]
ci:forward

**Comment by nhpatt** [comment]
ci:forward

**Comment by nhpatt** [comment]
ci:forward

**Comment by nhpatt** [comment]
ci:forward

**Comment by nhpatt** [comment]
ci:forward
